astropy/utils/metadata/core.py

```diff
@@ -100,18 +149,25 @@ def __get__(self, instance, owner):
         return instance._meta
 
     def __set__(self, instance, value):
-        # The 'default' value is `None`, but we want to set it to an empty `OrderedDict`
+        # The 'default' value is `None`, but we want to set it to an empty `Mapping`
         # if it is `None` so that we can always assume it is a `Mapping` and not have
         # to check for `None` everywhere.
         if value is None:
-            instance._meta = self.default_factory()
+            self._set_on_instance(instance, self.default_factory())
```

Comment:
Looking at this, I feel that because of my mutterings, this has become too complicated...  Since `object.__setattr__` seems to be the standard for (frozen) dataclasses, how about just
```
if value is None:
    value = self.default_factory()
elif not ...
    ...

try:
    instance._meta = value
except AttributeError:
    if is_dataclass(instance) [and is_frozen??]:
        object.__setattr__(instance, "_meta", value)
    else:
        raise
```


And then get rid of the possible `obj_setattr`, etc.

Reply:
Done! 
The ``is_dataclass(instance)`` check is so fast I avoided the `try-except`.

Reply:
``is_dataclass`` is just ``hasattr(instance, '__dataclass_fields__')``.